[SPARK-31959][SQL][3.0] Fix Gregorian-Julian micros rebasing while switching standard time zone offset

MaxGekk opened the pull request:
### What changes were proposed in this pull request?
Fix the bug in microseconds rebasing during transitions from one standard time zone offset to another one. In the PR, I propose to change the implementation of `rebaseGregorianToJulianMicros` which performs rebasing via local timestamps. In the case of overlapping:
1. Check that the original instant belongs to earlier or later instant of overlapped local timestamp.
2. If it is an earlier instant, take zone and DST offsets from the previous day otherwise
3. Set time zone offsets to Julian timestamp from the next day.

Note: The fix assumes that transitions cannot happen more often than once per 2 days.

Adopt the test "SPARK-31959: JST -> HKT at Asia/Hong_Kong in 1945" to outdated tzdb. Old JDK can have outdated time zone database in which Asia/Hong_Kong doesn't have timestamp overlapping in 1945 at all. 

### Why are the changes needed?
1. Current implementation handles timestamps overlapping only during daylight saving time but overlapping can happen also during transition from one standard time zone to another one. For example in the case of `Asia/Hong_Kong`, the time zone switched from `Japan Standard Time` (UTC+9) to `Hong Kong Time` (UTC+8) on _Sunday, 18 November, 1945 01:59:59 AM_. The changes allow to handle the special case as well.
2. To fix the test failures on old JDK w/ outdated tzdb like on Jenkins machine `research-jenkins-worker-09`.

### Does this PR introduce _any_ user-facing change?
It might affect micros rebasing in before common era when not-optimised version of `rebaseGregorianToJulianMicros()` is used directly.

### How was this patch tested?
1. By existing tests in `DateTimeUtilsSuite`, `RebaseDateTimeSuite`, `DateFunctionsSuite`, `DateExpressionsSuite` and `TimestampFormatterSuite`.
2. Added new test to `RebaseDateTimeSuite`
3. Regenerated `gregorian-julian-rebase-micros.json` with the step of 30 minutes, and got the same JSON file. The JSON file isn't affected because previously it was generated with the step of 1 week. And the spike in diffs/switch points during 1 hour of timestamp overlapping wasn't detected.

Authored-by: Max Gekk <max.gekk@gmail.com>
Signed-off-by: Wenchen Fan <wenchen@databricks.com>
(cherry picked from commit c259844)
Signed-off-by: Dongjoon Hyun <dongjoon@apache.org>
(cherry picked from commit eae1747)
Signed-off-by: Max Gekk <max.gekk@gmail.com>

MaxGekk committed:
[SPARK-31959][SQL][TEST-JAVA11] Fix Gregorian-Julian micros rebasing while switching standard time zone offset

Fix the bug in microseconds rebasing during transitions from one standard time zone offset to another one. In the PR, I propose to change the implementation of `rebaseGregorianToJulianMicros` which performs rebasing via local timestamps. In the case of overlapping:
1. Check that the original instant belongs to earlier or later instant of overlapped local timestamp.
2. If it is an earlier instant, take zone and DST offsets from the previous day otherwise
3. Set time zone offsets to Julian timestamp from the next day.

Note: The fix assumes that transitions cannot happen more often than once per 2 days.

Current implementation handles timestamps overlapping only during daylight saving time but overlapping can happen also during transition from one standard time zone to another one. For example in the case of `Asia/Hong_Kong`, the time zone switched from `Japan Standard Time` (UTC+9) to `Hong Kong Time` (UTC+8) on _Sunday, 18 November, 1945 01:59:59 AM_. The changes allow to handle the special case as well.

It might affect micros rebasing in before common era when not-optimised version of `rebaseGregorianToJulianMicros()` is used directly.

1. By existing tests in `DateTimeUtilsSuite`, `RebaseDateTimeSuite`, `DateFunctionsSuite`, `DateExpressionsSuite` and `TimestampFormatterSuite`.
2. Added new test to `RebaseDateTimeSuite`
3. Regenerated `gregorian-julian-rebase-micros.json` with the step of 30 minutes, and got the same JSON file. The JSON file isn't affected because previously it was generated with the step of 1 week. And the spike in diffs/switch points during 1 hour of timestamp overlapping wasn't detected.

Closes #28787 from MaxGekk/HongKong-tz-1945.

Authored-by: Max Gekk <max.gekk@gmail.com>
Signed-off-by: Wenchen Fan <wenchen@databricks.com>
(cherry picked from commit c259844df8b6690b752a1c67b241de2981cdb5fe)
Signed-off-by: Max Gekk <max.gekk@gmail.com>

MaxGekk commented:
@cloud-fan Please, take a look at the backport of https://github.com/apache/spark/pull/28787 to branch-3.0

SparkQA commented:
**[Test build #123906 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/123906/testReport)** for PR 28809 at commit [`35245c0`](https://github.com/apache/spark/commit/35245c0fb5e018ca6c1d4b1b9a7dda36ae5cdc4b).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/28530/
Test PASSed.

cloud-fan commented:
do you know what caused the conflicts?

MaxGekk commented:
> do you know what caused the conflicts?

As usual - `fromMillis()`

SparkQA commented:
**[Test build #123906 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/123906/testReport)** for PR 28809 at commit [`35245c0`](https://github.com/apache/spark/commit/35245c0fb5e018ca6c1d4b1b9a7dda36ae5cdc4b).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/123906/
Test FAILed.

MaxGekk commented:
Jenkins, retest this, please

SparkQA commented:
**[Test build #123925 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/123925/testReport)** for PR 28809 at commit [`35245c0`](https://github.com/apache/spark/commit/35245c0fb5e018ca6c1d4b1b9a7dda36ae5cdc4b).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/28547/
Test PASSed.

gatorsmile commented:
We need to split such a long Jenkin job to multiple shorter Jenkins Jobs

SparkQA commented:
**[Test build #123925 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/123925/testReport)** for PR 28809 at commit [`35245c0`](https://github.com/apache/spark/commit/35245c0fb5e018ca6c1d4b1b9a7dda36ae5cdc4b).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/123925/
Test PASSed.

dongjoon-hyun commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/util/RebaseDateTimeSuite.scala:
```diff
@@ -409,4 +409,40 @@ class RebaseDateTimeSuite extends SparkFunSuite with Matchers with SQLHelper {
     }
   }
+
+  test("SPARK-31959: JST -> HKT at Asia/Hong_Kong in 1945") {
```
Please hold on this. This test case is flaky in `master` branch already.
- https://amplab.cs.berkeley.edu/jenkins/view/Spark%20QA%20Test%20(Dashboard)/job/spark-master-test-maven-hadoop-3.2-hive-2.3/lastCompletedBuild/testReport/org.apache.spark.sql.catalyst.util/RebaseDateTimeSuite/SPARK_31959__JST____HKT_at_Asia_Hong_Kong_in_1945/
```
- SPARK-31959: JST -> HKT at Asia/Hong_Kong in 1945 *** FAILED ***
  -766744200000000 did not equal -766747800000000 (RebaseDateTimeSuite.scala:432)
```

dongjoon-hyun reviewed (changes requested):
After some investigations, I found that the failures seem to reported consistently on `research-jenkins-worker-09`. It might be Amplap Jenkins host issue (Java version or environment). We had better fix the host or make this PR more robust on those problems before merging this PR to `branch-3.0`.

MaxGekk commented:
> It might be Amplap Jenkins host issue (Java version or environment).

It uses JDK w/ outdated time zone database (not clear from log which version):
```
JAVA_HOME=/usr/lib/jvm/java-8-oracle
```
other jenkins machines have:
```
JAVA_HOME=/usr/java/jdk1.8.0_191
```
If we are not able to upgrade JDK 1.8 to the recent version, can we have at least the same JDK on all jenkins machines?

MaxGekk commented:
I am going to skip the test checks if JDK tzdb is outdated and Asia/Hong_Kong doesn't have timestamps overlapping in 1945 at all.

MaxGekk commented:
I will cherry-pick https://github.com/apache/spark/pull/28832 here when it will be merged to master.

MaxGekk committed:
[SPARK-31959][SQL][TESTS][FOLLOWUP] Adopt the test "SPARK-31959: JST -> HKT at Asia/Hong_Kong in 1945" to outdated tzdb

Old JDK can have outdated time zone database in which `Asia/Hong_Kong` doesn't have timestamp overlapping in 1946 at all. This PR changes the test "SPARK-31959: JST -> HKT at Asia/Hong_Kong in 1945" in `RebaseDateTimeSuite`, and makes it tolerant to the case.

To fix the test failures on old JDK w/ outdated tzdb like on Jenkins machine `research-jenkins-worker-09`.

No

By running the test on old JDK

Closes #28832 from MaxGekk/HongKong-tz-1945-followup.

Authored-by: Max Gekk <max.gekk@gmail.com>
Signed-off-by: Dongjoon Hyun <dongjoon@apache.org>
(cherry picked from commit eae1747b663adaa612681d799dcaa11915cd1692)
Signed-off-by: Max Gekk <max.gekk@gmail.com>

dongjoon-hyun commented:
It's merged now. Thanks!

SparkQA commented:
**[Test build #124060 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/124060/testReport)** for PR 28809 at commit [`63898cf`](https://github.com/apache/spark/commit/63898cf439762d198a022e0fe7e4c4a4bb196350).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/28679/
Test PASSed.

SparkQA commented:
**[Test build #124060 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/124060/testReport)** for PR 28809 at commit [`63898cf`](https://github.com/apache/spark/commit/63898cf439762d198a022e0fe7e4c4a4bb196350).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/124060/
Test PASSed.

cloud-fan commented:
thanks, merging to 3.0!
